### DISCUSSION TITLE
Model sources

### DISCUSSION THREAD

**Discussion by iboB** [alpaca-core/alpaca-core-sdk#61] [Dev]
How will model sources work?

I guess the most important question is how they will update their specific model manifests. Some options, and their implications, here include:

* Pull. A source is periodically updated from the associated server (or other medium)
	* The model source manager (provider) will have to do these pulls and it's not yet clear how to do it. Simply adding a timer seems wasteful as most apps would either not use this feature or only query models once at the start.
	* the implementation needs to be a polymorphic object for different mediums
* Push. A source relies on the remote medium to notify it of asset manifest changes
	* It will be really unpleasant to have the manifest stored in a simple http file server. Since the server can't push the information a hidden timer-based pull will need to be implemented per model source. Again this is wasteful and adds complexity to source implementations
 	* the implementation needs to be a polymorphic object for different mediums
* Immutable. A source is immutable. It obtains its model manifest on construction and it remains the same forever. 
    * Updating model manifests happens by removing outdated sources and adding new ones in their place
    * Think of a model source not as `http://ac.com/models.json`, but as `http://ac.com/models-12.53.3243.json`
    * Model manifests can be assets (accessible by multiple processes)
    * Updates are transferred to the user
    * The implementation is trivial. Any tweaks are left entirely up to the user

I'm heavily leaning to the last approach. It seems most future proof. If we ever decide to help with updating model manifests, we can do it externally (as library plugins, rather than api changes)

**Comment by pminev** [comment]
`1)` and `3)` seems pretty similar to me with the difference that with 3) everything is in the hands of the user. I'm leaning to `3)` too since we are giving more freedom to the users and the implementation as you say must be trivial. Furthermore, we really can implement something which they can use it as a basis.

**Comment by IvanFilipov** [comment]
`3)` is the most convenient to me, also!

**Comment by tzanko-matev** [comment]
Another vote for 3
